myhdl/_fixbv.py

```diff
@@ -0,0 +1,517 @@
+        self._max = 2**(nrbits-1)
+        fxval = self._from_float(val)
+        intbv.__init__(self, fxval, min=self._min, max=self._max)
+        #self._fval = val
+
+        if self._nrbits != nrbits:
+            errstr = "ERROR: intbv number of bits != fixbv number of bits %d,%d" \
+                     % (self._nrbits, nrbits)
+            raise ValueError(errstr)
+
+        # make sure things were setup ok
+        self._handleBounds()
+        
+        # round mode and overflow mode
+        if round_mode is None:
+            self._round_mode = 'floor'
+        else:
+            self._round_mode = round_mode
+        
+        if overflow_mode is None:
+            self._overflow_mode = 'saturate'
+        else:
+            self._overflow_mode = overflow_mode
+
+
+    def _handleBounds(self):
+        """ check bounds """
+        intbv._handleBounds(self)
+
+    ###################################################################### 
+    # properties
+    ######################################################################
+    @property
+    def format(self):
+        return tuple(self._W[:])
+
+    @property
+    def res(self):
+        _res = 2.0 ** (-1*self._W._fwl)
+        return _res
+
+    @property
+    def max(self):
+        return self.__max
+
+    @property
+    def min(self):
+        return self.__min
+        
+    ###################################################################### 
+    # overloaded functions
+    ######################################################################
+    def __copy__(self):
+        min, max, res = self._minmaxres()
+        retval = fixbv(self._to_float(), min, max, res)
+                       #round_mode=self.round_mode,
+                       #overflow_mode=self.overflow_mode)
+        return retval
+    
+    def __deepcopy__(self, visit):
+        min, max, res = self._minmaxres()
+        retval = fixbv(self._to_float(), min, max, res)
+                       #round_mode=self.round_mode,
+                       #overflow_mode=self.overflow_mode)
+        return retval
+
+    def __getitem__(self,key):
+        if isinstance(key,(tuple,list)):
+            fwl = key[2] if len(key) == 3 else key[0]-key[1]-1
+            res = 2**-(fwl)
+            nmax = 2**(key[1])
+            nmin = -nmax
+            if self._val is None:
+                fval = self._ifval
+            else:
+                fval = self._to_float()
+            return fixbv(fval, min=nmin, max=nmax, res=res)
+        else:
+            # @todo: check for negative index and convert
+            #        to the underlying intbv indexes
+            slc = intbv(self._val, _nrbits=self._nrbits)
+            return slc.__getitem__(key)
+
+    def __setitem__(self, key, val):
+        if isinstance(val, fixbv):
+            assert self._W._fwl == val._W._fwl
+            v = val._val
+        else:
+            v = val
+        # @todo: convert negative keys to the correct bit index
```

Comment:
Need to add the round and alignment functions here, the majority of the work for this project is the automatic alignment and sizing.